misc: SRNoC model

kunpai opened the pull request:
This is a model of a statically-scheduled circuit-switched NoC for superconductors.
Modeled off the SRNoC paper by Michelogiannakis et al.

kunpai committed:
misc: SRNoC model

This is a model of a statically-scheduled circuit-switched
NoC for superconductors.
Modeled off the SRNoC paper by Michelogiannakis et al.

powerjg commented on src/network/DataCell.cc:
All files need copyright.

powerjg reviewed (changes requested):
I see almost no comments. You need to add comments to this code. 

Feel free to use AI to help you, but you have to have comments.

kunpai committed:
misc: Add comments and license

Comments for better code understandability added.
License added.

kunpai committed:
misc: Add test script to configs

Test Script for SRNoC model

kunpai committed:
misc: Encoding for zero in SRNoC

One additional time slot for zero.
Also, event added to show when a packet is delivered within
a connection window.

kunpai committed:
misc: Add more traffic patterns

Hotspot and All-to-All

Not integrated as selectable options yet

kunpai committed:
misc: Update SRNoC model

Move hardware constants out of model into Python script.
Delays are now model parameters.
Refactor formula stats of power and area out.

kunpai committed:
misc: Update SRNoC stats

Packets per window is a histogram.
Also, minor change to Param description
in SuperNetwork to align with default.

kunpai committed:
misc: Fix exitSimLoop bug

Fixed a bug where the simulation exited before
the last packet was delivered.

kunpai committed:
misc: Effect of Frequency on SRNoC

Frequency now affects time slot and
connection window.

kunpai committed:
misc: Refactor out precomputed schedule

Now, destination is computed in the network event
If it is allowed, send it.
If not, push it into the buffer and send when
allowed.

kunpai committed:
misc: SuperNetwork functionality moved to Layer

Layer send packets based on scheduling
Network now has a vector of Layers
Config script also changed to reflect this change

kunpai committed:
misc: Minor style changes

Removed some unused commented code
Changed year on license from 2022 to 2025

kunpai committed:
misc: Miscellaneous changes to SRNoC Model

DPRINTF output to show layer number.
Multiple dynamic ranges, which dictate number
of layers.
Payload Specific delay had a small bug in it.

kunpai committed:
misc: Infinite packets implemented

If max packets is -1, then the model generates and delivers infinite packets

kunpai committed:
misc: Fix bug related to multiple layers

Multiple layers did not follow maximum packets
Fixed it by making exit simulation at the network level
Layers inform the network when they are done processing
When all layers reach maximum packets, exit

kunpai committed:
misc: Move delay parameters into Layer

Crosspoint delay, merger delay etc. are now part of
a Layer.

kunpai committed:
misc: Implement ability to choose traffic mode

Random and hotspot are currently supported.
Implemented via enums and exposed function
in Layer to set the traffic mode.
Edited main SRNoC config script to accept traffic
mode as well.

kunpai committed:
misc: Implement file-based schedule & other misc changes

Providing a file input loads a schedule from that file.
Fixed a bug where the last packet was not fully delivered
before exiting the simulation.
Rename .cc and .hh files to match gem5 specifications.
Changed some local variables to snake case as well.
Removed some unused commented code.

kunpai committed:
misc: Fix argument parsers in SRNoC config script

Made sure all arguments are being parsed correctly.
Also, fixed if __name__=='__main__' to
'__m5_main__'.

kunpai committed:
misc: Code clean-up

Cleaned code; removed unused functions.
Renamed variables to follow gem5 naming conventions.

kunpai committed:
misc: Add hold time

According to SRNoC paper, a connection window is
setup time + RL time slots + hold time

Updated scheduling to reflect the above.

kunpai committed:
misc: Improve timing model

Timing was made more accurate based on the SRNoC paper

kunpai committed:
misc: Add All-to-All traffic mode

One DataCell schedules to all other DataCells

kunpai committed:
misc: Add tornado traffic pattern

Stresses bisection bandwidth

kunpai committed:
misc: Dynamic histogram bins

kunpai committed:
misc: Change timing delays and add new stat

Stat tracks the missed packets per cell

kunpai committed:
misc: Small change in SConscript

Apostrophe style changed for consistency

kunpai committed:
misc: Move data generation to scheduler

Random payload generation within range of values possible

kunpai committed:
misc: Add shuffling to all-to-all traffic

kunpai committed:
misc: Update SRNoC params

Dynamic range updated to RL Time Slots to reflect paper semantics
Also, RL Time Slots also dictate maximum packets sent in a window,
implemented that behavior

kunpai committed:
misc: Refactor out # of RL Time Slots as configurable

Parameters like delays of components like crosspoint,
splitter, merger, etc. are configurable.
The above decide the length of a time slot.
According to the paper, 8 time slots of 92 ps, make
the frequency of the network 1.4 GHz. This makes sense,
as 736 ps time period is roughly 1.4 GHz.
So, given configurable frequencies, the Layer computes
the max range of values and the max packets sent
per window.
Multiple frequencies can be defined to make multiple
Layers.

kunpai committed:
misc: Rename DataCell to BufferedPort

DataCell is no longer an accurate name,
since data is generated by the scheduler.

kunpai committed:
misc: Add some fatals

If timeSlot is greater than clockPeriod, fatal because no values
can be sent on that Layer

kunpai committed:
misc: Payload specific delay formula changed

More accurate; shows correct behavior from
the paper as well

kunpai committed:
misc: Add bit complement and nearest neighbor traffic patterns

They were in the SRNoC paper.

kunpai committed:
misc: Multiple packets from 1 src to 1 dest in 1 window

Paper indicates we can send more than 1 value in a window
between a src and a dest. Only caveat being they need to
be DISTINCT values.

kunpai committed:
misc: Align area and power model with ground truth

Closer to values seen in paper

kunpai committed:
misc: Minor changes

Add stat to track the attempted communications and latency
Fix bug with maximum communications possible in a window

kunpai committed:
misc: Replace packet with value

Since communications in SRNoC do not have headers & flits,
values is the more accurate term.

kunpai committed:
misc: Replace packet with value in config script

kunpai committed:
misc: Add latency tracking, no buffer mode

Latency tracking (in buffer + network delay)
No buffer mode; does not store in a buffer at all
Config script accepts a --no-buffer arg

kunpai committed:
misc: Add param to target messages per second, bug fixes

Bug was with sending multiple values from one src to a
dest. Fixed it to cap at number of rlTimeSlots in a
connection window.
Using the above concept, we can reach a target number
of messages per second.

kunpai committed:
misc: Add buffer depth

This is done as a param to control the injection rate
for nearest neighbor and all-to-all

kunpai committed:
misc: Configurable buffer depth

Removed no buffer mode bool and collapsed its functionality onto the
buffer depth param

kunpai committed:
misc: Merge SRNoC and refactor buffer depth

Resolve merge conflicts after pulling origin/SRNoC.
Collapses noBufferMode into bufferDepth=0 for un-buffered mode and
updates configs/supernetwork/SRNoC.py, Layer.py, layer.hh, layer.cc.

kunpai committed:
misc: Fixing bugs

kunpai committed:
misc: Add active sources

Based on a parameter, we can either set a raw number or a fraction of the srcs
to be active FOR SOURCE SENDING.
Destinations are all still fully active.
There is also a boolean that can determine if these active sources should be
sequential or if they can be random.
Also, right now I am using a hard-coded seed to make sure the random sources
are the same.
Perhaps might consider setting seed through a parameter later.

kunpai committed:
misc: Move seq/rand active ports bool

Move it out of mutually exclusive parser
